File: Plugins/AWS/src/android/com/shortsands/aws/s3/AwsS3.java
package com.shortsands.aws.s3;

/**
 * Created by garygriswold on 5/19/17.
 */
import android.content.Context;
import android.util.Log;

import com.amazonaws.HttpMethod;
import com.amazonaws.services.s3.model.GeneratePresignedUrlRequest;

import com.amazonaws.mobileconnectors.s3.transferutility.TransferObserver;
import com.amazonaws.mobileconnectors.s3.transferutility.TransferUtility;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3Client;

import java.io.File;
import java.net.URL;
import java.util.Date;

public class AwsS3 {

    static String TAG = "AwsS3";
    static String VIDEO_ANALYTICS_BUCKET = "video.analytics.shortsands";

    TransferUtility transferUtility;

    public AwsS3(Context context) {
        super();
        AmazonS3 s3 = new AmazonS3Client(Credentials.AWS_BIBLE_APP);
        this.transferUtility = new TransferUtility(s3, context);
        // Region is not used here, why is it required in iOS can it be left out
    }
    /////////////////////////////////////////////////////////////////////////
    // URL signing Functions
    /////////////////////////////////////////////////////////////////////////
    /**
     * This method produces a presigned URL for a GET from an AWS S3 bucket
     */
    public URL preSignedUrlGET(String s3Bucket, String s3Key, int expires) {
        GeneratePresignedUrlRequest request =
            new GeneratePresignedUrlRequest(s3Bucket, s3Key, HttpMethod.GET);
        Date expiration = new Date(new Date().getTime() + expires);
        request.withExpiration(expiration);

        AmazonS3 s3client = new AmazonS3Client(Credentials.AWS_BIBLE_APP);
        URL url = s3client.generatePresignedUrl(request);
        Log.d(TAG, url.toExternalForm());
        return url;
    }
    /**
     * This method produces a presigned URL for a PUT to an AWS S3 bucket
     */
    public URL preSignedUrlPUT(String s3Bucket, String s3Key, int expires, String contentType) {
        GeneratePresignedUrlRequest request =
                new GeneratePresignedUrlRequest(s3Bucket, s3Key, HttpMethod.PUT);
        Date expiration = new Date(new Date().getTime() + expires);
        request.withExpiration(expiration);
        request.withContentType(contentType);

        AmazonS3 s3client = new AmazonS3Client(Credentials.AWS_BIBLE_APP);
        URL url = s3client.generatePresignedUrl(request);
        Log.d(TAG, url.toExternalForm());
        return url;
    }
    /////////////////////////////////////////////////////////////////////////
    // Download Functions
    /////////////////////////////////////////////////////////////////////////
    /**
     * Download Text to String object
     * Is there any reason why this is different than binary.  Do I need to handle encoding?
     */
    public void downloadText(String s3Bucket, String s3Key, DownloadTextListener listener) {
        try {
            File tempFile = File.createTempFile("downloadText", null);
            Log.d(TAG, "temp file created " + tempFile.getAbsolutePath());
            listener.setFile(tempFile);
            TransferObserver observer = this.transferUtility.download(s3Bucket, s3Key, tempFile);
            observer.setTransferListener(listener);
        } catch(Exception err) {
            Log.e(TAG, "Error in downloadText " + s3Bucket + "." + s3Key + "  " + err.toString());
        }
    }
    /**
     * Download Binary object to Data, receiving code might need to convert it needed form
     */
    public void downloadData(String s3Bucket, String s3Key, DownloadDataListener listener) {
        File tempFile = null;
        try {
            tempFile = File.createTempFile("downloadData", null);
            listener.setFile(tempFile);
            TransferObserver observer = this.transferUtility.download(s3Bucket, s3Key, tempFile);
            observer.setTransferListener(listener);
        } catch(Exception err) {
            Log.e(TAG, "Error in downloadData " + s3Bucket + "." + s3Key + "  " + err.toString());
        }
    }
    /**
     * Download File.  This works for binary and text files.
     */
    public void downloadFile(String s3Bucket, String s3Key, File file, DownloadFileListener listener) {
        TransferObserver observer = this.transferUtility.download(s3Bucket, s3Key, file);
        observer.setTransferListener(listener);
    }
    /**
     * Download zip file that contains one file, unzips and extracts file.
     * If there is ever a need to download and unzip an archive a separate method should be written
     * rather than modify this one.
     */
    public void downloadZipFile(String s3Bucket, String s3Key, File file, DownloadZipFileListener listener) {
        File zipFile = null;
        try {
	        listener.setFile(file);
            zipFile = File.createTempFile("downloadZip", "");
            listener.setZipFile(zipFile);
            TransferObserver observer = this.transferUtility.download(s3Bucket, s3Key, zipFile);
            observer.setTransferListener(listener);
        } catch (Exception err) {
            Log.e(TAG, "Error in downloadZipFile " + err.toString());
        }
    }
    /////////////////////////////////////////////////////////////////////////
    // Upload Functions
    /////////////////////////////////////////////////////////////////////////
    /**
     * Upload Analytics in Text form, such as JSON to analytics bucket
     */
    public void uploadVideoAnalytics(String sessionId, String timestamp, String data, UploadDataListener listener) {
        String s3Key = sessionId + timestamp;
        uploadText(AwsS3.VIDEO_ANALYTICS_BUCKET, s3Key, data, listener);
    }
    /**
     * Upload string object to bucket
     */
    public void uploadText(String s3Bucket, String s3Key, String data, UploadDataListener listener) {
        File tempFile = null;
        try {
            tempFile = File.createTempFile("uploadText", "");
            listener.setFile(tempFile);
            FileManager.writeTextFully(tempFile, data);
            TransferObserver observer = this.transferUtility.upload(s3Bucket, s3Key, tempFile);
            observer.setTransferListener(listener);
        } catch(Exception err) {
            Log.e(TAG, "Error in uploadText " + err.toString());
        }
    }
    /**
     * Upload object in Data form to bucket.  Data must be prepared to correct form
     * before calling this function.
     */
    public void uploadData(String s3Bucket, String s3Key, byte[] data, UploadDataListener listener) {
        File tempFile = null;
        try {
            tempFile = File.createTempFile("uploadData", "");
            listener.setFile(tempFile);
            FileManager.writeBinaryFully(tempFile, data);
            TransferObserver observer = this.transferUtility.upload(s3Bucket, s3Key, tempFile);
            observer.setTransferListener(listener);
        } catch(Exception err) {
            Log.e(TAG, "Error in uploadData " + err.toString());
        }
    }
    /**
     * Upload file to bucket, this works for text or binary files
     */
    public void uploadFile(String s3Bucket, String s3Key, File file, UploadFileListener listener) {
        if (file.exists() && file.isFile()) {
            TransferObserver observer = this.transferUtility.upload(s3Bucket, s3Key, file);
            observer.setTransferListener(listener);
        } else {
            Log.e(TAG, "Error: File does not exist: " + file.getAbsolutePath());
        }
    }
}

File: Plugins/AWS/src/ios/AwsS3.swift
//
//  AwsS3.swift
//  AWS_S3_Prototype
//
//  Created by Gary Griswold on 5/15/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

import Foundation
//import AWSS3

public class AwsS3 {
    
    static let VIDEO_ANALYTICS_BUCKET = "video.analytics.shortsands"
    
    private var transfer: AWSS3TransferUtility
    
    init() {
        let configuration = AWSServiceConfiguration(region: AWSRegionType.USWest2,
                                                    credentialsProvider: Credentials.AWS_BIBLE_APP)
        AWSServiceManager.default().defaultServiceConfiguration = configuration
        //AWSS3TransferUtility.interceptApplication was not set, because we do not need it.
        self.transfer = AWSS3TransferUtility.default()
    }
    /**
    * A Unit Test Method
    */
    func echo3(message: String) -> String {
	    return message;
    }
    /////////////////////////////////////////////////////////////////////////
    // URL signing Functions
    /////////////////////////////////////////////////////////////////////////
    /**
    * This method produces a presigned URL for a GET from an AWS S3 bucket
    */
    func preSignedUrlGET(s3Bucket: String, s3Key: String, expires: Int,
                         complete: @escaping (_ url:URL?) -> Void) {
        let request = AWSS3GetPreSignedURLRequest()
        request.httpMethod = AWSHTTPMethod.GET
        request.bucket = s3Bucket
        request.key = s3Key
        request.expires = Date(timeIntervalSinceNow: TimeInterval(expires))
        presignURL(request: request, complete: complete)
    }
    /**
    * This method produces a presigned URL for a PUT to an AWS S3 bucket
    */
    func preSignedUrlPUT(s3Bucket: String, s3Key: String, expires: Int, contentType: String,
                         complete: @escaping (_ url:URL?) -> Void) {
        let request = AWSS3GetPreSignedURLRequest()
        request.httpMethod = AWSHTTPMethod.PUT
        request.bucket = s3Bucket
        request.key = s3Key
        request.expires = Date(timeIntervalSinceNow: TimeInterval(expires))
        request.contentType = contentType
        presignURL(request: request, complete: complete)
    }
    /**
    * This private method is the actual preSignedURL call
    */
    private func presignURL(request: AWSS3GetPreSignedURLRequest,
                            complete: @escaping (_ url:URL?) -> Void) {
        let builder = AWSS3PreSignedURLBuilder.default()
        builder.getPreSignedURL(request).continueWith { (task) -> URL? in
            if let error = task.error {
                print("Error: \(error)")
                complete(nil)
            } else {
                let presignedURL = task.result!
                complete(presignedURL as URL)
            }
            return nil
        }
    }
    /////////////////////////////////////////////////////////////////////////
    // Zip Functions
    /////////////////////////////////////////////////////////////////////////
    /**
     * Zip Utility for use on files.
     */
    func zip(sourceFile: String, targetDir: String) -> Bool {
        //let done = SSZipArchive.createZipFile(atPath: sourceFile, withContentsOfDirectory: targetDir)
        //(BOOL)createZipFileAtPath:(NSString *)path withContentsOfDirectory:(NSString *)directoryPath;
        let done = false
        return done
    }
    /**
     * UnZip Utility for use on files.
     */
    func unzip(sourceFile: String, targetDir: String) -> Bool {
        //let done = SSZipArchive.unzipFile(atPath: sourceFile, toDestination: targetDir)
        //print("DID unzip \(done)")
        let done = false
        return done
    }
    /////////////////////////////////////////////////////////////////////////
    // Download Functions
    /////////////////////////////////////////////////////////////////////////
    /**
    * Download Text to String object
    */
    func downloadText(s3Bucket: String, s3Key: String,
                      complete: @escaping (_ error:Error?, _ data:String?) -> Void) {
        let completionHandler: AWSS3TransferUtilityDownloadCompletionHandlerBlock = {(task, url, data, error) -> Void in
            DispatchQueue.main.async(execute: {
                if let err = error {
                    print("ERROR in s3.downloadText \(s3Bucket) \(s3Key) Error: \(err)")
                } else {
                    print("SUCCESS in s3.downloadText \(s3Bucket) \(s3Key)")
                }
                let datastring = (data != nil) ? String(data: data!, encoding: String.Encoding.utf8) as String? : ""
                complete(error, datastring)
            })
        }
        download(s3Bucket: s3Bucket, s3Key: s3Key, completionHandler: completionHandler)
    }
    /**
    * Download Binary object to Data, receiving code might need to convert it needed form
    */
    func downloadData(s3Bucket: String, s3Key: String,
                      complete: @escaping (_ error:Error?, _ data:Data?) -> Void) {
        let completionHandler: AWSS3TransferUtilityDownloadCompletionHandlerBlock = {(task, url, data, error) -> Void in
            DispatchQueue.main.async(execute: {
                if let err = error {
                    print("ERROR in s3.downloadData \(s3Bucket) \(s3Key) Error: \(err)")
                } else {
                    print("SUCCESS in s3.downloadData \(s3Bucket) \(s3Key)")
                }
                complete(error, data)
            })
        }
        download(s3Bucket: s3Bucket, s3Key: s3Key, completionHandler: completionHandler)
    } 
    /**
    * Download File.  This works for binary and text files. This method does not use
    * TransferUtility.fileDownload, because it is unable to provide accurate errors
    * when the file IO fails.
    */
    func downloadFile(s3Bucket: String, s3Key: String, filePath: URL,
                      complete: @escaping (_ error:Error?) -> Void) {
        let completionHandler: AWSS3TransferUtilityDownloadCompletionHandlerBlock = {(task, url, data, error) -> Void in
            DispatchQueue.main.async(execute: {
                if let err = error {
                    print("ERROR in s3.downloadFile \(s3Bucket) \(s3Key) Error: \(err)")
                    complete(error)
                } else {
                    do {
                        try data?.write(to: filePath, options: Data.WritingOptions.atomic)
                        print("SUCCESS in s3.downloadFile \(s3Bucket) \(s3Key)")
                        complete(nil)
                    } catch {
                        print("File IO Error in s3.downloadFile \(s3Bucket) \(s3Key)")
                        complete(nil) // pass error here, if I knew how to catch or create one.
                    }
                }
            })
        }
        download(s3Bucket: s3Bucket, s3Key: s3Key, completionHandler: completionHandler)
    }   
    /**
    * Download zip file and unzip it.  Like Download File this does not use 
    * TransferUtility.fileDownload because its error reporting is poor.
    *
    * NOTE: This is NOT working for files that were zipped using MacOs version of PKzip
    * and so it should be be used in the App.!!!!! May 19, 2017 GNG
    */
    func downloadZipFile(s3Bucket: String, s3Key: String, filePath: URL,
                         complete: @escaping (_ error:Error?) -> Void) {
	    downloadFile(s3Bucket: s3Bucket, s3Key: s3Key, filePath: filePath, complete: complete)
	    /*                     
        let completionHandler: AWSS3TransferUtilityDownloadCompletionHandlerBlock = {(task, url, data, error) -> Void in
            DispatchQueue.main.async(execute: {
                if let err = error {
                    print("ERROR in s3.downloadZipFile \(s3Bucket) \(s3Key) Error: \(err)")
                    complete(error)
                } else if (data != nil) {
                    let tempURL = URL(fileURLWithPath: NSTemporaryDirectory() + "/" + NSUUID().uuidString)
                    do {
                        try data!.write(to: tempURL, options: Data.WritingOptions.atomic)
                        let unzipped = self.unzip(sourceFile: tempURL.absoluteString,
                                                  targetDir: tempURL.absoluteString)
                        print("Unzip succeeded? \(unzipped)")
                        // locate file in the tempURL directory
                        // move the file to filePath
                        let destUrl = URL(fileURLWithPath: NSHomeDirectory() + filePath)
                        let fileManager = FileManager.default
                        try fileManager.moveItem(at: tempURL, to: destUrl)
                        print("SUCCESS in s3.downloadZipFile \(s3Bucket) \(s3Key)")
                        complete(nil)
                    } catch {
                        print("File IO Error in s3.downloadZipFile \(s3Bucket) \(s3Key)")
                        complete(nil) // pass error here, if I knew how to catch or create one.
                    }
                } else {
                    print("NO ERROR in s3.downloadZipFile and NO DATA \(s3Bucket) \(s3Key)") // add error if poss
                    complete(nil)
                }
            })
        }
        download(s3Bucket: s3Bucket, s3Key: s3Key, completionHandler: completionHandler)
        */
    }
    /**
    * Internal getObject method
    */
    private func download(s3Bucket: String, s3Key: String,
                          completionHandler: @escaping AWSS3TransferUtilityDownloadCompletionHandlerBlock) {
        self.transfer.downloadData(
            fromBucket: s3Bucket,
            key: s3Key,
            expression: nil,
            completionHandler: completionHandler)
        //.continueWith has been dropped, because it did not report errors
    }
    /////////////////////////////////////////////////////////////////////////
    // Upload Functions
    /////////////////////////////////////////////////////////////////////////
    /**
    * Upload Analytics in Text form, such as JSON to analytics bucket
    */
    func uploadVideoAnalytics(sessionId: String, timestamp: String, data: String,
                         complete: @escaping (_ error: Error?) -> Void) {
        let s3Key = sessionId + timestamp
        let textData = data.data(using: String.Encoding.utf8)
        uploadData(s3Bucket: AwsS3.VIDEO_ANALYTICS_BUCKET, s3Key: s3Key, data: textData!, contentType: "text/plain",
                    complete: complete)
    }
    /**
    * Upload string object to bucket
    */
    func uploadText(s3Bucket: String, s3Key: String, data: String,
                      complete: @escaping (_ error: Error?) -> Void) {
        let textData = data.data(using: String.Encoding.utf8)
        uploadData(s3Bucket: s3Bucket, s3Key: s3Key, data: textData!, contentType: "text/plain", complete: complete)
    }
    /**
     * Upload object in Data form to bucket.  Data must be prepared to correct form
     * before calling this function.
     */
    func uploadData(s3Bucket: String, s3Key: String, data: Data, contentType: String,
                    complete: @escaping (_ error: Error?) -> Void) {
        let completionHandler: AWSS3TransferUtilityUploadCompletionHandlerBlock = {(task, error) -> Void in
            DispatchQueue.main.async(execute: {
                if let err = error {
                    print("ERROR in s3.uploadData \(s3Bucket) \(s3Key) Error: \(err)")
                } else {
                    print("SUCCESS in s3.uploadData \(s3Bucket) \(s3Key)")
                }
                complete(error)
            })
        }
        self.transfer.uploadData(data,
                            bucket: s3Bucket,
                            key: s3Key,
                            contentType: contentType,
                            expression: nil,
                            completionHandler: completionHandler)
        //.continueWith has been dropped, because it did not report errors
    }
    /**
    * Upload file to bucket, this works for text or binary files
    */
    func uploadFile(s3Bucket: String, s3Key: String, filePath: URL, contentType: String,
                    complete: @escaping (_ error: Error?) -> Void) {
        do {
            // Uncertain what .uncached means and if entry is best choice.
            let data = try Data(contentsOf: filePath, options: Data.ReadingOptions.uncached)
            uploadData(s3Bucket: s3Bucket, s3Key: s3Key, data: data, contentType: contentType,
                       complete: complete)
        } catch {
            // Can I capture error here, and use it in message and complete
            print("ERROR in s3.uploadFile, while reading file \(s3Bucket) \(s3Key)")
            complete(nil)
        }
    }
}
